superduperdb/container/vector_index.py

```diff
@@ -98,8 +99,9 @@ def on_load(self, db: DB) -> None:
                 ),
                 create=True,  # type: ignore[call-arg]
             )
+        clt = self.indexing_listener.select.collection  # type: ignore[attr-defined]
```

Comment:
So now we know how to handle these to make it clearer what's going on!!

----

Look at the definition of `self.indexing_listener`.  it has two possible types, `str | Listener`.  But only one, `Listener` has the `.select` member!  

The type checker _rightly_ complains that in some possible codepaths, `self.indexing_listener` is a string.

The solution is to make sure that we never get to that line unless `self.indexing_listener` is a `Listener`.

So you put this line first: `assert isinstance(self.indexing_listener, Listener)`

Now, this might reveal another issue, but it will probably be the same thing where you then need to add ``assert isinstance(self.indexing_listener.select, ...)`.

These asserts have several purposes!

* They fix the mypy problem
* They document the potential issue at that point
* If by some horrible chance the value _is_ wrong, you get a better error
* We can see and count them in order to see where we should get rid of "color changing variables"


 "Color changing variables" are variables like this which start off with one type and get converted to some other type by the time they get used.  You can see that they add a risk.

Reply:
@rec the type error is not regarding self.indexing_listener.select 
we already have casted it to Listener, look at 4-5 lines above.

The error is that self.indexing_listener.select has no attribute `collection`

Reply:
<img width="1426" alt="image" src="https://github.com/SuperDuperDB/superduperdb/assets/138251983/f957f391-895c-4b79-81cf-58cf5ee0c3da">

Reply:
Oops, so sorry to have missed this. 

The issue is because this code "knows" that this is in fact a MongoDb query, and has a `collection` field, but the base query type `Select` doesn't have this.

We do a lot of "declaring an object as the base class and using the derived class" and of course that has an element of risk to it, and worse, looks bad in people's IDEs.

For the moment, try 

`assert hasattr(self.indexing_listener.select, 'collection')`

We should probably extract a common base class...

Reply:
This will work while typing but during execution for some queries 
like limit

some 

```python
VectorIndex(indexing_listerning=Listener(select=Collection(name='x').find().limit(100) ))

```
It will give error!

@blythed Let me know your thoughts here,